design/design-generic-references.md

```diff
@@ -0,0 +1,319 @@
+
+* be configurable at runtime
+* use the existing [Kubernetes API object reference] convention 
+  as much as possible
+* provide alternative convention in cases where [Kubernetes API object reference]
+  convention is not suitable
+* be implemented without or requiring only light changes in RBAC
+* support referring multiple objects (namespaced and cluster-scoped) and their
+  fields
+* support value transformation before assignment
+
+## Proposal
+
+We would like to introduce a new cluster-scoped `Referable` (alternative name
+could be `InjectableValue`) type whose instances declare values that can be
+assigned to an object field:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: Referable
+metadata:
+  name: referable-vpc-id
+spec:
+  objects:
+    - apiVersion: ec2.aws.crossplane.io/v1alpha1
+      kind: VPC
+      name: main-vpc
+      fieldPath: status.atProvider.id
+```
+
+The syntax of the `fieldPath` field is similar to what we
+use in Composition, in line with the [Kubernetes API Conventions on field selection].
+
+After deploying the above instance to the cluster, its reconciliation gets 
+triggered and once the referred object exists and the field is set, 
+the referable value is emitted in the object status part:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: Referable
+metadata:
+  name: referable-vpc-id
+spec:
+  objects:
+    - apiVersion: ec2.aws.crossplane.io/v1alpha1
+      kind: VPC
+      name: main-vpc
+      fieldPath: status.atProvider.id
+status:
+  value: foo-vpc-id # the value found in the requested field
+  conditions:
+    - type: Ready
+      status: True
+      reason: Available
+```
+
+Now, the value can be referred in the usual way within a managed resource:
+
+```yaml
+spec:
+  vpcIdRef: referable-vpc-id
+```
+
+The reference resolver could first try to find the `referable-vpc-id` `VPC` instance, 
+and fallback to `referable-vpc-id` `Referable` if the former does not exist. 
+If the instance is ready, the value is read from `status.value` and 
+set to `spec.vpcId` of the managed resource. Of course, if needed, and if it 
+makes more sense, the resolver could first look for the existence of `Referable` 
+object first. With that strategy, we would be able to overwrite the referencing
+mechanism set at the compile-time.
+
+Referring to a namespaced object would be possible as well:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: Referable
+metadata:
+  name: region
+spec:
+  objects:
+    - apiVersion: v1
+      kind: ConfigMap
+      name: common-settings
+      namespace: crossplane-system
+      fieldPath: data.region
+```
+
+or we could match an object with a label selector if its name is unknown or not
+static:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: Referable
+metadata:
+  name: referable-vpc-id
+spec:
+  objects:
+    - apiVersion: ec2.aws.crossplane.io/v1alpha1
+      kind: VPC
+      matchingLabels:
```

Comment:
what happens if multiple objects are selected with labels? Reminds me of https://github.com/crossplane/crossplane/pull/3981 when we thought about similar quite late in the implementation